File: sdk/dart/lib/src/fetch.dart
/// The network boundary (SPEC.md section 3.2).
///
/// An interface rather than direct `dart:io` calls, so the orchestration in
/// `updater.dart` can be tested without a server, and so a host that must route
/// through its own proxy or client certificate can substitute its own.
library;

import 'dart:async';
import 'dart:io';
import 'dart:math' as math;
import 'dart:typed_data';

class FetchException implements Exception {
  FetchException(this.url, this.message, {this.statusCode, this.retryable});

  final Uri url;
  final String message;
  final int? statusCode;

  /// When null, callers infer from [statusCode] / message (network vs 4xx).
  final bool? retryable;

  bool get isRetryable {
    if (retryable != null) return retryable!;
    final code = statusCode;
    if (code == null) return true; // connection / timeout
    if (code == 408 || code == 429) return true;
    if (code >= 500 && code <= 599) return true;
    return false;
  }

  @override
  String toString() => 'FetchException($url): $message';
}

/// Progress of an artifact download, including a smoothed throughput estimate.
class DownloadProgress {
  const DownloadProgress({
    required this.received,
    required this.total,
    required this.bytesPerSecond,
    this.eta,
  });

  final int received;
  final int total;
  final double bytesPerSecond;
  final Duration? eta;

  double? get fraction => total > 0 ? received / total : null;
}

/// Called as bytes arrive. Throughput is a short sliding window, not lifetime average.
typedef ProgressCallback = void Function(DownloadProgress progress);

/// Result of probing a URL for size and Range support.
class ResourceProbe {
  const ResourceProbe({
    required this.acceptsRanges,
    this.contentLength,
  });

  final bool acceptsRanges;
  final int? contentLength;
}

/// Shared helper: sliding-window bytes/sec from incremental [onBytes] reports.
class ThroughputMeter {
  ThroughputMeter({this.window = const Duration(seconds: 1)});

  final Duration window;
  final List<({DateTime at, int bytes})> _samples = [];
  int _received = 0;

  int get received => _received;

  DownloadProgress observe({required int total, int? delta}) {
    if (delta != null && delta > 0) {
      _received += delta;
      final now = DateTime.now();
      _samples.add((at: now, bytes: delta));
      final cutoff = now.subtract(window);
      while (_samples.isNotEmpty && _samples.first.at.isBefore(cutoff)) {
        _samples.removeAt(0);
      }
    }
    final spanMs = _samples.isEmpty
        ? 0
        : math.max(
            1,
            _samples.last.at.difference(_samples.first.at).inMilliseconds,
          );
    final windowBytes =
        _samples.fold<int>(0, (sum, s) => sum + s.bytes);
    // If only one sample, use elapsed since that sample against wall clock.
    final bps = _samples.isEmpty
        ? 0.0
        : _samples.length == 1
            ? windowBytes /
                math.max(
                  0.001,
                  DateTime.now().difference(_samples.first.at).inMilliseconds /
                      1000.0,
                )
            : windowBytes / (spanMs / 1000.0);

    Duration? eta;
    if (bps > 1 && total > _received) {
      final secs = (total - _received) / bps;
      eta = Duration(milliseconds: (secs * 1000).round());
    }
    return DownloadProgress(
      received: _received,
      total: total,
      bytesPerSecond: bps,
      eta: eta,
    );
  }

  void seed(int alreadyReceived) {
    _received = alreadyReceived;
  }
}

abstract class Fetcher {
  /// Fetches a small document whole.
  Future<Uint8List> getBytes(Uri url, {required Duration timeout});

  /// Probes whether the resource supports byte ranges (and optional length).
  Future<ResourceProbe> probe(Uri url, {required Duration timeout});

  /// Streams a large file to disk (single connection).
  ///
  /// When [startOffset] > 0, sends `Range: bytes=startOffset-` and appends to
  /// [destination]. A `200` response means the server ignored Range: the
  /// caller should truncate and restart (this method throws
  /// [FetchException] with message `range not honored`).
  Future<void> download(
    Uri url,
    File destination, {
    required Duration idleTimeout,
    int startOffset = 0,
    int? knownTotal,
    ProgressCallback? onProgress,
  });

  /// Fetches inclusive byte range `[start, endInclusive]` into [destination]
  /// as a contiguous file starting at offset 0 (length = end-start+1).
  /// Requires HTTP 206.
  Future<void> downloadRange(
    Uri url, {
    required File destination,
    required int start,
    required int endInclusive,
    required Duration idleTimeout,
    void Function(int bytesJustReceived)? onBytes,
  });

  void close();
}

/// The default [Fetcher], on `dart:io`.
class HttpFetcher implements Fetcher {
  HttpFetcher({String userAgent = 'rup-client/2'})
      : _client = HttpClient()..userAgent = userAgent;

  final HttpClient _client;

  /// Redirect limit from SPEC.md section 3.2. Object storage and release hosts
  /// redirect routinely, so refusing to follow any would break real
  /// deployments; allowing unbounded ones invites a loop.
  static const maxRedirects = 5;

  @override
  Future<Uint8List> getBytes(Uri url, {required Duration timeout}) async {
    final response = await _open(url, timeout: timeout);
    try {
      final builder = BytesBuilder(copy: false);
      await for (final chunk in response.timeout(timeout)) {
        builder.add(chunk);
      }
      return builder.takeBytes();
    } on TimeoutException {
      throw FetchException(url, 'timed out after ${timeout.inSeconds}s');
    }
  }

  @override
  Future<ResourceProbe> probe(Uri url, {required Duration timeout}) async {
    // Prefer HEAD. Some CDNs mishandle HEAD; fall back to a one-byte Range GET.
    try {
      final response = await _open(
        url,
        timeout: timeout,
        method: 'HEAD',
      );
      await response.drain<void>();
      final accepts = _acceptsBytesRanges(response);
      final length = response.contentLength >= 0 ? response.contentLength : null;
      return ResourceProbe(acceptsRanges: accepts, contentLength: length);
    } on FetchException catch (error) {
      if (error.statusCode == 405 || error.statusCode == 501) {
        // Fall through to Range probe.
      } else if (error.statusCode != null &&
          error.statusCode! >= 400 &&
          error.statusCode! < 500) {
        rethrow;
      }
    }

    final response = await _open(
      url,
      timeout: timeout,
      headers: {HttpHeaders.rangeHeader: 'bytes=0-0'},
    );
    try {
      if (response.statusCode == 206) {
        final total = _totalFromContentRange(response);
        await response.drain<void>();
        return ResourceProbe(acceptsRanges: true, contentLength: total);
      }
      if (response.statusCode == 200) {
        final length =
            response.contentLength >= 0 ? response.contentLength : null;
        await response.drain<void>();
        return ResourceProbe(acceptsRanges: false, contentLength: length);
      }
      await response.drain<void>();
      throw FetchException(
        url,
        'HTTP ${response.statusCode}',
        statusCode: response.statusCode,
      );
    } finally {
      // drained above
    }
  }

  @override
  Future<void> download(
    Uri url,
    File destination, {
    required Duration idleTimeout,
    int startOffset = 0,
    int? knownTotal,
    ProgressCallback? onProgress,
  }) async {
    final headers = <String, String>{};
    if (startOffset > 0) {
      headers[HttpHeaders.rangeHeader] = 'bytes=$startOffset-';
    }

    final response = await _open(
      url,
      timeout: idleTimeout,
      headers: headers.isEmpty ? null : headers,
    );

    if (startOffset > 0) {
      if (response.statusCode == 200) {
        await response.drain<void>();
        throw FetchException(
          url,
          'range not honored',
          statusCode: 200,
          retryable: false,
        );
      }
      if (response.statusCode != 206) {
        await response.drain<void>();
        throw FetchException(
          url,
          'HTTP ${response.statusCode}',
          statusCode: response.statusCode,
        );
      }
    }

    final contentLength =
        response.contentLength >= 0 ? response.contentLength : null;
    final total = knownTotal ??
        (startOffset > 0 && contentLength != null
            ? startOffset + contentLength
            : contentLength);

    await destination.parent.create(recursive: true);
    final sink = destination.openWrite(
      mode: startOffset > 0 ? FileMode.append : FileMode.write,
    );
    final meter = ThroughputMeter()..seed(startOffset);
    try {
      await for (final chunk in response.timeout(idleTimeout)) {
        sink.add(chunk);
        if (onProgress != null) {
          onProgress(meter.observe(total: total ?? 0, delta: chunk.length));
        } else {
          meter.observe(total: total ?? 0, delta: chunk.length);
        }
      }
      await sink.flush();
    } on TimeoutException {
      await sink.close();
      throw FetchException(
        url,
        'stalled for ${idleTimeout.inSeconds}s after ${meter.received} bytes',
      );
    } catch (error) {
      await sink.close();
      if (error is FetchException) rethrow;
      throw FetchException(url, 'download failed: $error');
    }
    await sink.close();
  }

  @override
  Future<void> downloadRange(
    Uri url, {
    required File destination,
    required int start,
    required int endInclusive,
    required Duration idleTimeout,
    void Function(int bytesJustReceived)? onBytes,
  }) async {
    if (endInclusive < start) {
      throw ArgumentError('endInclusive < start');
    }
    final expected = endInclusive - start + 1;
    final response = await _open(
      url,
      timeout: idleTimeout,
      headers: {HttpHeaders.rangeHeader: 'bytes=$start-$endInclusive'},
    );

    if (response.statusCode == 200) {
      await response.drain<void>();
      throw FetchException(
        url,
        'range not honored',
        statusCode: 200,
        retryable: false,
      );
    }
    if (response.statusCode != 206) {
      await response.drain<void>();
      throw FetchException(
        url,
        'HTTP ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }

    await destination.parent.create(recursive: true);
    final sink = destination.openWrite(mode: FileMode.write);
    var got = 0;
    try {
      await for (final chunk in response.timeout(idleTimeout)) {
        sink.add(chunk);
        got += chunk.length;
        onBytes?.call(chunk.length);
      }
      await sink.flush();
    } on TimeoutException {
      await sink.close();
      throw FetchException(
        url,
        'stalled for ${idleTimeout.inSeconds}s after $got bytes of range',
      );
    } catch (error) {
      await sink.close();
      if (error is FetchException) rethrow;
      throw FetchException(url, 'range download failed: $error');
    }
    await sink.close();

    if (got != expected) {
      throw FetchException(
        url,
        'range size mismatch: got $got want $expected',
        retryable: true,
      );
    }
  }

  Future<HttpClientResponse> _open(
    Uri url, {
    required Duration timeout,
    String method = 'GET',
    Map<String, String>? headers,
  }) async {
    if (url.scheme != 'http' && url.scheme != 'https') {
      throw FetchException(url, 'unsupported scheme "${url.scheme}"');
    }

    Future<HttpClientResponse> send(Uri target) async {
      try {
        final request = await _client.openUrl(method, target).timeout(timeout);
        request.followRedirects = false;
        headers?.forEach(request.headers.set);
        return await request.close().timeout(timeout);
      } on TimeoutException {
        throw FetchException(target, 'timed out after ${timeout.inSeconds}s');
      } on SocketException catch (error) {
        throw FetchException(target, 'connection failed: ${error.message}');
      }
    }

    var response = await send(url);
    var current = url;
    for (var hop = 0; hop < maxRedirects; hop++) {
      if (!_isRedirect(response.statusCode)) break;

      final location = response.headers.value(HttpHeaders.locationHeader);
      if (location == null) {
        throw FetchException(current, 'redirect without a Location header',
            statusCode: response.statusCode);
      }
      final next = current.resolve(location);

      if (current.scheme == 'https' && next.scheme != 'https') {
        throw FetchException(
            current, 'refusing redirect from https to ${next.scheme}');
      }

      await response.drain<void>();
      response = await send(next);
      current = next;
    }

    if (_isRedirect(response.statusCode)) {
      throw FetchException(current, 'more than $maxRedirects redirects',
          statusCode: response.statusCode);
    }

    // For HEAD/GET with Range we accept 206 as success at the call site.
    // Here only reject obvious error statuses so probe/download can inspect 206/200.
    if (response.statusCode >= 400) {
      final code = response.statusCode;
      await response.drain<void>();
      throw FetchException(current, 'HTTP $code', statusCode: code);
    }
    return response;
  }

  static bool _acceptsBytesRanges(HttpClientResponse response) {
    final raw = response.headers.value(HttpHeaders.acceptRangesHeader);
    if (raw == null) return false;
    return raw.toLowerCase().split(',').any((p) => p.trim() == 'bytes');
  }

  static int? _totalFromContentRange(HttpClientResponse response) {
    final cr = response.headers.value(HttpHeaders.contentRangeHeader);
    if (cr == null) return null;
    final slash = cr.lastIndexOf('/');
    if (slash < 0 || slash + 1 >= cr.length) return null;
    return int.tryParse(cr.substring(slash + 1));
  }

  static bool _isRedirect(int status) =>
      status == 301 ||
      status == 302 ||
      status == 303 ||
      status == 307 ||
      status == 308;

  @override
  void close() => _client.close(force: true);
}

File: sdk/node/src/fetch.ts
/**
 * The network boundary (SPEC.md section 3.2).
 *
 * An interface rather than direct `fetch` calls, so the orchestration in
 * `updater.ts` can be tested without a server, and so a host that must route
 * through its own proxy or client certificate can substitute its own.
 */

import { createHash } from "node:crypto";
import { createReadStream, createWriteStream } from "node:fs";
import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";

export class FetchError extends Error {
  readonly url: string;
  readonly statusCode?: number;
  private readonly retryable?: boolean;

  constructor(
    url: string | URL,
    message: string,
    options: { statusCode?: number; retryable?: boolean } = {},
  ) {
    super(`FetchError(${String(url)}): ${message}`);
    this.name = "FetchError";
    this.url = String(url);
    this.detail = message;
    this.statusCode = options.statusCode;
    this.retryable = options.retryable;
  }

  readonly detail: string;

  get isRetryable(): boolean {
    if (this.retryable !== undefined) return this.retryable;
    const code = this.statusCode;
    if (code === undefined) return true; // connection / timeout
    if (code === 408 || code === 429) return true;
    return code >= 500 && code <= 599;
  }
}

/** Progress of an artifact download, including a smoothed throughput estimate. */
export interface DownloadProgress {
  received: number;
  total: number;
  bytesPerSecond: number;
  etaMs?: number;
  fraction?: number;
}

export type ProgressCallback = (progress: DownloadProgress) => void;

/** Result of probing a URL for size and Range support. */
export interface ResourceProbe {
  acceptsRanges: boolean;
  contentLength?: number;
}

/** Sliding-window bytes/sec from incremental byte reports. */
export class ThroughputMeter {
  private readonly samples: { at: number; bytes: number }[] = [];
  private receivedBytes = 0;

  constructor(private readonly windowMs = 1000) {}

  get received(): number {
    return this.receivedBytes;
  }

  seed(alreadyReceived: number): void {
    this.receivedBytes = alreadyReceived;
  }

  observe(total: number, delta?: number): DownloadProgress {
    if (delta !== undefined && delta > 0) {
      this.receivedBytes += delta;
      const now = Date.now();
      this.samples.push({ at: now, bytes: delta });
      const cutoff = now - this.windowMs;
      while (this.samples.length > 0 && this.samples[0]!.at < cutoff) {
        this.samples.shift();
      }
    }

    let bps = 0;
    if (this.samples.length === 1) {
      const elapsed = Math.max(1, Date.now() - this.samples[0]!.at);
      bps = (this.samples[0]!.bytes / elapsed) * 1000;
    } else if (this.samples.length > 1) {
      const spanMs = Math.max(
        1,
        this.samples[this.samples.length - 1]!.at - this.samples[0]!.at,
      );
      const windowBytes = this.samples.reduce((sum, s) => sum + s.bytes, 0);
      bps = (windowBytes / spanMs) * 1000;
    }

    const progress: DownloadProgress = {
      received: this.receivedBytes,
      total,
      bytesPerSecond: bps,
    };
    if (total > 0) progress.fraction = this.receivedBytes / total;
    if (bps > 1 && total > this.receivedBytes) {
      progress.etaMs = Math.round(((total - this.receivedBytes) / bps) * 1000);
    }
    return progress;
  }
}

export interface Fetcher {
  /** Fetches a small document whole. */
  getBytes(url: URL, timeoutMs: number): Promise<Uint8Array>;

  /** Probes whether the resource supports byte ranges (and optional length). */
  probe(url: URL, timeoutMs: number): Promise<ResourceProbe>;

  /**
   * Streams a large file to disk (single connection). When `startOffset > 0`,
   * sends `Range: bytes=startOffset-` and appends. A `200` response means the
   * server ignored Range: this throws `FetchError` with detail
   * `range not honored` so the caller can truncate and restart.
   */
  download(
    url: URL,
    destination: string,
    options: {
      idleTimeoutMs: number;
      startOffset?: number;
      knownTotal?: number;
      onProgress?: ProgressCallback;
    },
  ): Promise<void>;

  /**
   * Fetches inclusive byte range `[start, endInclusive]` into `destination` as a
   * contiguous file starting at offset 0. Requires HTTP 206.
   */
  downloadRange(
    url: URL,
    options: {
      destination: string;
      start: number;
      endInclusive: number;
      idleTimeoutMs: number;
      onBytes?: (bytesJustReceived: number) => void;
    },
  ): Promise<void>;

  close(): void;
}

const REDIRECT_STATUSES = new Set([301, 302, 303, 307, 308]);

/**
 * Redirect limit from SPEC.md section 3.2. Object storage and release hosts
 * redirect routinely, so refusing to follow any would break real deployments;
 * allowing unbounded ones invites a loop.
 */
export const MAX_REDIRECTS = 5;

/**
 * Appends a cache-busting query parameter for mutable documents (SPEC §3.2).
 *
 * Immutable objects (manifest, artifact) must NOT be passed through this: they
 * are content-addressed, and busting their cache only removes the CDN benefit
 * that makes a release survive its own launch day.
 */
export function cacheBust(url: URL): URL {
  const out = new URL(url.toString());
  out.searchParams.set("t", String(Math.floor(Date.now() / 1000)));
  return out;
}

/** The default `Fetcher`, on global `fetch` (undici). */
export class HttpFetcher implements Fetcher {
  private readonly userAgent: string;

  constructor(options: { userAgent?: string } = {}) {
    this.userAgent = options.userAgent ?? "rup-client/2";
  }

  async getBytes(url: URL, timeoutMs: number): Promise<Uint8Array> {
    const response = await this.open(url, {
      timeoutMs,
      headers: { "cache-control": "no-cache" },
    });
    try {
      const buffer = await withTimeout(
        response.arrayBuffer(),
        timeoutMs,
        () => new FetchError(url, `timed out after ${timeoutMs}ms`),
      );
      return new Uint8Array(buffer);
    } catch (error) {
      if (error instanceof FetchError) throw error;
      throw new FetchError(url, `read failed: ${String(error)}`);
    }
  }

  async probe(url: URL, timeoutMs: number): Promise<ResourceProbe> {
    // Prefer HEAD. Some CDNs mishandle HEAD; fall back to a one-byte Range GET.
    try {
      const response = await this.open(url, { timeoutMs, method: "HEAD" });
      await drain(response);
      const length = parseLength(response.headers.get("content-length"));
      const result: ResourceProbe = {
        acceptsRanges: acceptsBytesRanges(response.headers.get("accept-ranges")),
      };
      if (length !== undefined) result.contentLength = length;
      return result;
    } catch (error) {
      if (!(error instanceof FetchError)) throw error;
      const status = error.statusCode;
      if (status !== 405 && status !== 501) {
        if (status !== undefined && status >= 400 && status < 500) throw error;
      }
    }

    const response = await this.open(url, {
      timeoutMs,
      headers: { range: "bytes=0-0" },
    });
    if (response.status === 206) {
      const total = totalFromContentRange(response.headers.get("content-range"));
      await drain(response);
      const result: ResourceProbe = { acceptsRanges: true };
      if (total !== undefined) result.contentLength = total;
      return result;
    }
    if (response.status === 200) {
      const length = parseLength(response.headers.get("content-length"));
      await drain(response);
      const result: ResourceProbe = { acceptsRanges: false };
      if (length !== undefined) result.contentLength = length;
      return result;
    }
    await drain(response);
    throw new FetchError(url, `HTTP ${response.status}`, {
      statusCode: response.status,
    });
  }

  async download(
    url: URL,
    destination: string,
    options: {
      idleTimeoutMs: number;
      startOffset?: number;
      knownTotal?: number;
      onProgress?: ProgressCallback;
    },
  ): Promise<void> {
    const startOffset = options.startOffset ?? 0;
    const headers: Record<string, string> = {};
    if (startOffset > 0) headers.range = `bytes=${startOffset}-`;

    const response = await this.open(url, {
      timeoutMs: options.idleTimeoutMs,
      headers,
    });

    if (startOffset > 0) {
      if (response.status === 200) {
        await drain(response);
        throw new FetchError(url, "range not honored", {
          statusCode: 200,
          retryable: false,
        });
      }
      if (response.status !== 206) {
        await drain(response);
        throw new FetchError(url, `HTTP ${response.status}`, {
          statusCode: response.status,
        });
      }
    }

    const contentLength = parseLength(response.headers.get("content-length"));
    const total =
      options.knownTotal ??
      (startOffset > 0 && contentLength !== undefined
        ? startOffset + contentLength
        : contentLength) ??
      0;

    await mkdir(dirname(destination), { recursive: true });
    const meter = new ThroughputMeter();
    meter.seed(startOffset);

    const sink = createWriteStream(destination, {
      flags: startOffset > 0 ? "a" : "w",
    });
    const source = toNodeStream(url, response, options.idleTimeoutMs, (n) => {
      const progress = meter.observe(total, n);
      options.onProgress?.(progress);
    });

    await pipeline(source, sink);
  }

  async downloadRange(
    url: URL,
    options: {
      destination: string;
      start: number;
      endInclusive: number;
      idleTimeoutMs: number;
      onBytes?: (bytesJustReceived: number) => void;
    },
  ): Promise<void> {
    const { start, endInclusive } = options;
    if (endInclusive < start) throw new RangeError("endInclusive < start");
    const expected = endInclusive - start + 1;

    const response = await this.open(url, {
      timeoutMs: options.idleTimeoutMs,
      headers: { range: `bytes=${start}-${endInclusive}` },
    });

    if (response.status === 200) {
      await drain(response);
      throw new FetchError(url, "range not honored", {
        statusCode: 200,
        retryable: false,
      });
    }
    if (response.status !== 206) {
      await drain(response);
      throw new FetchError(url, `HTTP ${response.status}`, {
        statusCode: response.status,
      });
    }

    await mkdir(dirname(options.destination), { recursive: true });
    let got = 0;
    const sink = createWriteStream(options.destination, { flags: "w" });
    const source = toNodeStream(url, response, options.idleTimeoutMs, (n) => {
      got += n;
      options.onBytes?.(n);
    });
    await pipeline(source, sink);

    if (got !== expected) {
      throw new FetchError(url, `range size mismatch: got ${got} want ${expected}`, {
        retryable: true,
      });
    }
  }

  close(): void {
    // Global fetch keeps its own pooled agent; nothing to release.
  }

  private async open(
    url: URL,
    options: {
      timeoutMs: number;
      method?: string;
      headers?: Record<string, string>;
    },
  ): Promise<Response> {
    if (url.protocol !== "http:" && url.protocol !== "https:") {
      throw new FetchError(url, `unsupported scheme "${url.protocol}"`);
    }

    const send = async (target: URL): Promise<Response> => {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), options.timeoutMs);
      try {
        return await fetch(target, {
          method: options.method ?? "GET",
          headers: { "user-agent": this.userAgent, ...options.headers },
          redirect: "manual",
          signal: controller.signal,
        });
      } catch (error) {
        if (controller.signal.aborted) {
          throw new FetchError(target, `timed out after ${options.timeoutMs}ms`);
        }
        throw new FetchError(target, `connection failed: ${String(error)}`);
      } finally {
        clearTimeout(timer);
      }
    };

    let current = url;
    let response = await send(current);

    for (let hop = 0; hop < MAX_REDIRECTS; hop++) {
      if (!REDIRECT_STATUSES.has(response.status)) break;

      const location = response.headers.get("location");
      if (location === null) {
        throw new FetchError(current, "redirect without a Location header", {
          statusCode: response.status,
        });
      }
      const next = new URL(location, current);
      if (current.protocol === "https:" && next.protocol !== "https:") {
        throw new FetchError(
          current,
          `refusing redirect from https to ${next.protocol}`,
        );
      }
      await drain(response);
      response = await send(next);
      current = next;
    }

    if (REDIRECT_STATUSES.has(response.status)) {
      throw new FetchError(current, `more than ${MAX_REDIRECTS} redirects`, {
        statusCode: response.status,
      });
    }

    if (response.status >= 400) {
      const status = response.status;
      await drain(response);
      throw new FetchError(current, `HTTP ${status}`, { statusCode: status });
    }
    return response;
  }
}

function toNodeStream(
  url: URL,
  response: Response,
  idleTimeoutMs: number,
  onBytes: (n: number) => void,
): Readable {
  const body = response.body;
  if (body === null) return Readable.from([]);

  const reader = body.getReader();
  return new Readable({
    async read() {
      try {
        const { done, value } = await withTimeout(
          reader.read(),
          idleTimeoutMs,
          () =>
            new FetchError(url, `stalled for ${idleTimeoutMs}ms`, {
              retryable: true,
            }),
        );
        if (done) {
          this.push(null);
          return;
        }
        onBytes(value.byteLength);
        this.push(Buffer.from(value));
      } catch (error) {
        this.destroy(
          error instanceof Error
            ? error
            : new FetchError(url, `read failed: ${String(error)}`),
        );
      }
    },
  });
}

async function withTimeout<T>(
  promise: Promise<T>,
  timeoutMs: number,
  onTimeout: () => Error,
): Promise<T> {
  let timer: NodeJS.Timeout | undefined;
  try {
    return await Promise.race([
      promise,
      new Promise<never>((_, reject) => {
        timer = setTimeout(() => reject(onTimeout()), timeoutMs);
      }),
    ]);
  } finally {
    if (timer !== undefined) clearTimeout(timer);
  }
}

async function drain(response: Response): Promise<void> {
  try {
    await response.body?.cancel();
  } catch {
    // Best effort.
  }
}

function acceptsBytesRanges(raw: string | null): boolean {
  if (raw === null) return false;
  return raw
    .toLowerCase()
    .split(",")
    .some((part) => part.trim() === "bytes");
}

function parseLength(raw: string | null): number | undefined {
  if (raw === null) return undefined;
  const value = Number.parseInt(raw, 10);
  return Number.isFinite(value) && value >= 0 ? value : undefined;
}

function totalFromContentRange(raw: string | null): number | undefined {
  if (raw === null) return undefined;
  const slash = raw.lastIndexOf("/");
  if (slash < 0 || slash + 1 >= raw.length) return undefined;
  const value = Number.parseInt(raw.slice(slash + 1), 10);
  return Number.isFinite(value) ? value : undefined;
}

/** Hashes bytes already in memory. For manifests, which are small by design. */
export function sha256OfBytes(bytes: Uint8Array): string {
  return createHash("sha256").update(bytes).digest("hex");
}

/** Streams the file through the hash rather than reading it whole. */
export async function sha256OfFile(path: string): Promise<string> {
  const hash = createHash("sha256");
  await pipeline(createReadStream(path), hash);
  return hash.digest("hex");
}
